proxy/src/metrics.rs

```diff
@@ -20,159 +29,203 @@ const DEFAULT_HTTP_REPORTING_TIMEOUT: Duration = Duration::from_secs(60);
 /// so while the project-id is unique across regions the whole pipeline will work correctly
 /// because we enrich the event with project_id in the control-plane endpoint.
 ///
-#[derive(Eq, Hash, PartialEq, Serialize, Debug)]
+#[derive(Eq, Hash, PartialEq, Serialize, Deserialize, Debug)]
 pub struct Ids {
     pub endpoint_id: String,
     pub branch_id: String,
 }
 
+impl Clone for Ids {
+    fn clone(&self) -> Self {
+        Self {
+            endpoint_id: self.endpoint_id.clone(),
+            branch_id: self.branch_id.clone(),
+        }
+    }
+    fn clone_from(&mut self, source: &Self) {
+        self.branch_id.clone_from(&source.branch_id);
+        self.endpoint_id.clone_from(&source.endpoint_id);
+    }
+}
+
+static STARTED_PROXY_COUNTERS: OnceLock<tokio::sync::mpsc::Sender<Arc<ProxyCounter>>> =
+    OnceLock::new();
+
 pub async fn task_main(config: &MetricCollectionConfig) -> anyhow::Result<Infallible> {
     info!("metrics collector config: {config:?}");
     scopeguard::defer! {
         info!("metrics collector has shut down");
     }
 
+    let (proxy_counters_tx, mut proxy_counters_rx) = mpsc::channel(256);
+    STARTED_PROXY_COUNTERS
+        .set(proxy_counters_tx)
+        .map_err(|_| anyhow::anyhow!("invalid proxy metrics state"))?;
+
     let http_client = http::new_client_with_timeout(DEFAULT_HTTP_REPORTING_TIMEOUT);
-    let mut cached_metrics: HashMap<Ids, (u64, DateTime<Utc>)> = HashMap::new();
-    let hostname = hostname::get()?.as_os_str().to_string_lossy().into_owned();
+    let mut proxy_counter_queue: VecDeque<Arc<ProxyCounter>> = VecDeque::new();
+
+    // fill with dummy data.
+    let mut events: Vec<Event<Ids>> = vec![
+        Event {
+            kind: EventType::Incremental {
+                start_time: Utc.timestamp_nanos(0),
+                stop_time: Utc.timestamp_nanos(0),
+            },
+            metric: "",
+            idempotency_key: String::new(),
+            value: 0,
+            extra: Ids {
+                endpoint_id: String::new(),
+                branch_id: String::new(),
+            },
+        };
+        CHUNK_SIZE
+    ];
+
+    let hostname = hostname::get()?;
+    let hostname = hostname.as_os_str().to_string_lossy();
 
     let mut ticker = tokio::time::interval(config.interval);
     loop {
         ticker.tick().await;
 
+        // acquire some new counters
+        for _ in 0..256 {
+            let Ok(counter) = proxy_counters_rx.try_recv() else { break };
+            proxy_counter_queue.push_back(counter);
+        }
```

Comment:
we could also flush the channel and wait for `ticker.tick()` in a `tokio::select!` loop so as soon as 256 OR ticker.tick() happen we would start pushing the updates? Didn't look at the channel implementation, but if it would happen that the channel is not emptied on any of the occasions, it might grow without a bound if we only act on the tick.

Reply:
Yeah that's probably a sensible idea. Only drain the channel while we're not running metrics.